libraries/networking/src/ByteSlice.cpp

```diff
@@ -0,0 +1,87 @@
+//
+//  Byteslice.cpp
+//  libraries/networking/src
+//
+//  Created by Heather Anderson on 2020-05-01.
+//  Copyright 2020 Vircadia.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#include "Byteslice.h"
+
+quint8 ByteSlice::_fallback{ 0 };
+
+ByteSlice::Bytestring::Bytestring(const quint8* source, size_t length) {
+    quint8* myContent = new quint8[length];
+    if (source != nullptr) {
+        memcpy(myContent, source, length);
+    } else {
+        memset(myContent, 0, length);
+    }
+    _content = myContent;
+    _fullLength = length;
+}
+
+ByteSlice::Bytestring::~Bytestring() {
+    delete[] _content;
+}
+
+quint8 ByteSlice::pop_front() {
+    if (!_length) {
+        return 0;
+    }
+    --_length;
+    return _content->_content[_offset++];
+}
+
+ByteSlice ByteSlice::substring(size_t offset, size_t length /* = static_cast<size_t>(-1) */) const {
+    if (offset >= _length) {
+        return ByteSlice();
+    }
+    size_t maxLen = _length - offset;
+    if (length == static_cast<size_t>(-1)) {
+        return ByteSlice(_content, _offset + offset, maxLen);
+    } else {
+        return ByteSlice(_content, _offset + offset, std::min(maxLen, length));
+    }
+}
+
+// create a new buffer and return a pointer to it
+void* ByteSlice::create(size_t length) {
+    _offset = 0;
+    _length = length;
+    _content = BytestringPointer::create(nullptr, length);
+    return const_cast<quint8*>(_content->_content);
+}
+
+// return a ByteSlice representing the concatenation of us and our argument.
+// In most cases will construct a new Bytestring, but will attempt to avoid doing so
+ByteSlice ByteSlice::concat(const ByteSlice& rhs) const {
+    // first of all check if either side is empty, if so we don't need to do anything
+    if (empty()) {
+        return rhs;
+    }
+    if (rhs.empty()) {
+        return *this;
+    }
+
+    // next check to see if we're "lucky" and we already have the data within ourselves
+    size_t rhsLength = rhs.length();
+    if (_offset + _length + rhsLength >= _content->_fullLength && memcmp(&_content->_content[_offset+_length], rhs.constData(), rhsLength) == 0) {
+        // "the answer was within you the whole time"
+        return ByteSlice(_content, _offset, _length + rhsLength);
+    }
+    if (rhs._offset >= _length && memcmp(&rhs._content->_content[rhs._offset - _length], constData(), _length) == 0) {
+        // "the answer was within *them* the whole time"
+        return ByteSlice(_content, rhs._offset - _length, _length + rhsLength);
+    }
```

Comment:
I'm not sure I understand the logic in these two if blocks...

the first one seems to be saying "if my size + rhs size is bigger than my full length, and the data after me matches rhs, make myself bigger to contain that data".  but if you're looking at data that's out of your bounds, even if the data happens to match rhs in value, there's no guarantee about what that other data is.  it could be used by something else and change out from under us, right?  I'm also confused about the logic for the second one.

but also, it seems like these memcmp checks would often be more expensive than just always allocating a new slice?

Reply:
okay looks like this particular code may need more checking of the comparisons, some of the inequality checks may be reversed.  This function was added late as the networking code needed some way to concatenate the contents of two packets together.  The memcmp can be skipped if it takes too long; I've been considering malloc() to be fairly intensive though (and in at least older heap implementations it would hold a global lock while allocating).

The idea was kinda if you had an original string named "the quick brown fox jumped over the lazy dog" and you had a ByteSlice representing "lazy" from that string and tried to concatenate " dog" to that then it would be able to reuse the ByteString and not have to allocate something new.  The arguments on this may be similar to that for inline functions though -- cost of passing all the checks (vs skipping once the first one fails) vs just going through and allocating a new object.

Depend on how other code works I may end up removing this function, if another (more specialized) class would be more appropriate (but then it's also a balance between complexity vs performance)

Reply:
I definitely see what this code is trying to accomplish, but yeah it seems like a tradeoff between how often we think that's going to actually happen and how expensive the checks are (I also assume the malloc is more expensive than the compare, but if we're doing both 90% of the time, is it really worth it?)

Reply:
I'm hoping the "if" statements outside of the call fail early the majority of the time so it never really gets to the memcmp() actually.